lib/src/auth/auth_controller.dart

```diff
@@ -1,34 +1,141 @@
+import 'dart:convert';
+
 import 'package:flutter/cupertino.dart';
-import 'package:flutter_secure_storage/flutter_secure_storage.dart';
+import 'package:http/http.dart' as http;
+import 'package:jwt_decoder/jwt_decoder.dart';
+
+import '../services/bedrock_service.dart';
+import 'auth_storage.dart';
 
-class AuthController extends ChangeNotifier {
-  final _baseApiUrl = const String.fromEnvironment('BASE_API');
-  final _apiVersion = '1';
-  final String _tokenKey = 'bedrock_token';
-  bool _authenticated = false;
+/// Interface used for declaring all the necessary methods for authentication
+/// and authorization.
+///
+/// This is also helpful for creating Mock tests to simulate API behavior
+abstract class IAuth {
+  Future<void> login(String email, String password);
+  Future<void> logout();
+  Future<void> register(
+    String email,
+    String firstName,
+    String lastName,
+    String password,
+  );
+  Future<void> resetPassword(String email);
+}
 
-  final FlutterSecureStorage storage = const FlutterSecureStorage();
+/// Responsible for interacting with Bedrock API through [BedrockService] and
+/// updating the UI by calling notifyListeners() to trigger a widget rebuild
+/// on authentication status updates.
+///
+/// [storage] Storage is a required class member which corresponds to
+/// [IAuthTokenStorage] for the mechanism to use for storing JWT tokens. This is
+/// mainly useful for providing other methods of storing token such as
+/// in-memory (for mock tests).
+class AuthController extends ChangeNotifier implements IAuth {
+  String? apiResponse;
+  final BedrockService apiService = BedrockService();
+  final IAuthTokenStorage storage;
+
+  AuthController({required this.storage});
 
   Future<bool> get authenticated async {
-    // String? token = await readAuthToken();
-    // return (token != null) ? JwtDecoder.isExpired(token) : false;
+    String? token = await storage.readAuthToken();
+    if (token != null) {
+      try {
+        return !JwtDecoder.isExpired(token);
+      } on FormatException {
+        return false;
+      }
+    } else {
+      return false;
+    }
+  }
 
-    return Future.delayed(Duration(seconds: 3), () => _authenticated);
+  @override
+  Future<void> login(String email, String password) async {
+    http.Response? response = await apiService.post('/auth/login', payload: {
+      'email': email,
+      'password': password,
+    });
+
+    if (response != null) {
+      var jsonResponse = jsonDecode(utf8.decode(response.bodyBytes)) as Map;
+      if (response.statusCode == 200) {
+        String token = jsonResponse['data']['token'];
+        if (!JwtDecoder.isExpired(token)) {
```

Comment:
I'm not quiet sure what you mean by having an interceptor, I'd appreciate it if you could provide me an example code snippet with more info about how you would refactor this piece of code or if you'd like, we can jump in a zoom meeting or something and just talk about it.

Reply:
You can find that code snippet on the url I sent you, but basically an interceptor is to intercept every call and make generic logic there.Here is a url with more information ( https://medium.com/flutter-community/dio-interceptors-in-flutter-17be4214f363)

Reply:
Thanks for sharing that blog post. I wasn't aware of the concept of interceptors until now so this is really interesting. I've gone ahead and replaced http with dio and added interceptors as you suggested. Please take a look at the updated code and let me know if it can be improved or optimized to work better.